inject/src/main/java/io/micronaut/context/DefaultBeanContext.java

```diff
@@ -4163,31 +4169,40 @@ public boolean isDefinitionEnabled(DefaultBeanContext context, @Nullable BeanRes
         }
 
         public <T> BeanDefinitionReference<T> getReference() {
-            if (reference == null || referenceEnabled == null || !referenceEnabled) {
+            // The reference needs to be assigned to a new variable as it can change between checks
+            Boolean refEnabled = referenceEnabled;
+            if (reference == null || refEnabled == null || !refEnabled) {
```

Comment:
should we also do   

`BeanDefinitionReference<?> ref = reference;`?

Reply:
It's only problematic when you read it twice: first null check then compare value, the second check might cause the NPE.